akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSupervisionSpec.scala

```diff
@@ -0,0 +1,656 @@
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+    }
+
+    "restart when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      restartingDecider))) {
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      resumingDecider))) {
+
+      pending // see blow FIXME
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne() // FIXME this throws IllegalStateException: Cannot pull while holding, only pushAndPull
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when onPull throws before pushing all buffered elements" in new TestSetup(Seq(
+      BufferingTestStage(resumingDecider))) {
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(1)
+      lastEvents() should be(Set(OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(2)))
+
+      downstream.requestOne()
+      // 3 => boom
+      // FIXME feels wrong? it should deliver OnNext(4)?
+      lastEvents() should be(Set(RequestOne))
```

Comment:
Thanks, this is interesting. Do you agree that for a restart we could support it, because the state is supposed to be cleared, and a new pull from upstream makes sense?

We can probably not support resume when onPull throws, and must resort to fail (stop).
Note that it is not many of the built in stages that have this problem. As far as I can see it is only `Expand`.

Just for my understanding. Let's say that this custom stage would emit an element downstream here as response to the exception instead of pulling. If a special Filter stage was installed afterwards to discard that kind of error elements everything would work? I'm somewhat confused of why we could not emulate something like that?